### PR TITLE
ccache: add 4.7.4

### PR DESCRIPTION
**Pull request by Cogitri**
Additionally, port to Conan v2 and update dependencies

---

- [X] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [X] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/ccache/all/conanfile.py [hunk 1]
```diff
@@ -1,9 +1,13 @@
-from conans import ConanFile, CMake, tools
-from conans.errors import ConanInvalidConfiguration
-import functools
+from conan import ConanFile
+from conan.tools.cmake import cmake_layout, CMake, CMakeToolchain, CMakeDeps
+from conan.tools.gnu import PkgConfigDeps
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.files import export_conandata_patches, apply_conandata_patches, copy, get
+from conan.tools.build import cross_building, check_min_cppstd
+from conan.tools.scm import Version
```
Comment (uilianries) [inline recipes/ccache/all/conanfile.py:7]:
```suggestion
from conan.tools.scm import Version
from conan.tools.microsoft import is_msvc
```

File: recipes/ccache/all/conanfile.py [hunk 2]
```diff
@@ -26,42 +30,48 @@ class CcacheConan(ConanFile):
...
 
-    generators = "cmake", "cmake_find_package_multi"
-
-    @property
-    def _source_subfolder(self):
-        return "source_subfolder"
-
     @property
     def _is_msvc(self):
         return str(self.settings.compiler) in ["Visual Studio", "msvc"]
```
Comment (uilianries) [inline recipes/ccache/all/conanfile.py:36]:
```suggestion
```
Replaced by https://docs.conan.io/en/latest/reference/conanfile/tools/microsoft.html#conan-tools-microsoft-is-msvc

File: recipes/ccache/all/conanfile.py [hunk 3]
```diff
@@ -79,35 +89,43 @@ def package_id(self):
         del self.info.settings.compiler
 
     def build_requirements(self):
-        if hasattr(self, "settings_build") and tools.cross_building(self) and \
+        self.build_requires("pkgconf/1.9.3")
+        if hasattr(self, "settings_build") and cross_building(self) and \
            self.settings.os == "Macos" and self.settings.arch == "armv8":
-            self.build_requires("cmake/3.22.0")
+            self.build_requires("cmake/3.25.1")
```
Comment (uilianries) [inline recipes/ccache/all/conanfile.py:95]:
```suggestion
            self.tool_requires("cmake/3.25.1")
```

File: recipes/ccache/all/conanfile.py [hunk 4]
```diff
@@ -79,35 +89,43 @@ def package_id(self):
         del self.info.settings.compiler
 
     def build_requirements(self):
-        if hasattr(self, "settings_build") and tools.cross_building(self) and \
+        self.build_requires("pkgconf/1.9.3")
```
Comment (uilianries) [inline recipes/ccache/all/conanfile.py:92]: Why need pkgconfig now?

File: recipes/ccache/all/conanfile.py [hunk 5]
```diff
@@ -1,9 +1,13 @@
...
+from conan.tools.cmake import cmake_layout, CMake, CMakeToolchain, CMakeDeps
+from conan.tools.gnu import PkgConfigDeps
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.files import export_conandata_patches, apply_conandata_patches, copy, get
+from conan.tools.build import cross_building, check_min_cppstd
+from conan.tools.scm import Version
 import os
 
-required_conan_version = ">=1.33.0"
+required_conan_version = ">=1.53.0"
```
Comment (uilianries) [inline recipes/ccache/all/conanfile.py:10]:
```suggestion
required_conan_version = ">=1.55.0"
```

https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmakedeps.html#set-property

File: recipes/ccache/all/conanfile.py [hunk 6]
```diff
@@ -79,35 +89,43 @@ def package_id(self):
...
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["REDIS_STORAGE_BACKEND"] = self.options.redis_storage_backend
+        tc.variables["HIREDIS_FROM_INTERNET"] = False
+        tc.variables["ENABLE_DOCUMENTATION"] = False
+        tc.variables["ENABLE_TESTING"] = False
+        tc.generate()
+        deps = PkgConfigDeps(self)
+        deps.generate()
+        if self._is_msvc:
```
Comment (uilianries) [inline recipes/ccache/all/conanfile.py:110]:
```suggestion
        if is_msvc(self):
```

File: recipes/ccache/all/conanfile.py [hunk 7]
```diff
@@ -79,35 +89,43 @@ def package_id(self):
...
+            deps = CMakeDeps(self)
+            deps.set_property("hiredislib", "cmake_target_name", "HIREDIS::HIREDIS")
```
Comment (uilianries) [inline recipes/ccache/all/conanfile.py:112]: Nice, much better than patches!

**Review by uilianries — COMMENTED** [review]
It's looking good, some few updates.

File: recipes/ccache/all/conanfile.py (hunk 4 above)
Comment (Cogitri) [reply recipes/ccache/all/conanfile.py:92]: For PkgConfigDeps - CMake aborts with an error if its not installed

**Comment by Cogitri** [comment]
Hm seems like `is_msvc` doesn't do the trick? 🤔  @uilianries

**Comment by Cogitri** [comment]
With pkgconf everything is detected correctly.

**Review by StellaSmith — APPROVED** [review]

File: recipes/ccache/all/conanfile.py [hunk 8]
```diff
@@ -26,42 +31,44 @@ class CcacheConan(ConanFile):
...
+                "clang": "9",
+                "apple-clang": "11",
+                "Visual Studio": "16.2",
+            }
+        else:
+            return {
+                "gcc": "6",
+                "clang": "6",
+                "apple-clang": "10",
+                "Visual Studio": "15.7" if Version(self.version) < "4.6" else "16.2",
```
Comment (prince-chrismc) [inline recipes/ccache/all/conanfile.py:55]: There's a new msvc compiler for 2.0, you can use this table to convert to the new versions

File: recipes/ccache/all/conanfile.py (hunk 4 above)
Comment (uilianries) [reply recipes/ccache/all/conanfile.py:91]: Please, use CMakeDeps instead, do not need pkgconfig.

File: recipes/ccache/all/conanfile.py [hunk 9]
```diff
@@ -79,35 +88,43 @@ def package_id(self):
...
+        tc.variables["ENABLE_DOCUMENTATION"] = False
+        tc.variables["ENABLE_TESTING"] = False
+        tc.generate()
+        deps = PkgConfigDeps(self)
+        deps.generate()
+        if is_msvc(self):
+            deps = CMakeDeps(self)
+            deps.set_property("hiredislib", "cmake_target_name", "HIREDIS::HIREDIS")
+            deps.set_property("zstd", "cmake_target_name", "ZSTD::ZSTD")
+            deps.generate()
```
Comment (uilianries) [inline recipes/ccache/all/conanfile.py:113]:
```suggestion
        deps = CMakeDeps(self)
        deps.set_property("hiredislib", "cmake_target_name", "HIREDIS::HIREDIS")
        deps.set_property("zstd", "cmake_target_name", "ZSTD::ZSTD")
        deps.generate()
```
Use CMakeDeps only, it should be enough for any platform.

File: recipes/ccache/all/conanfile.py [hunk 10]
```diff
@@ -79,35 +88,43 @@ def package_id(self):
         del self.info.settings.compiler
 
     def build_requirements(self):
-        if hasattr(self, "settings_build") and tools.cross_building(self) and \
+        self.build_requires("pkgconf/1.9.3")
+        if hasattr(self, "settings_build") and cross_building(self) and \
```
Comment (uilianries) [inline recipes/ccache/all/conanfile.py:92]: Why only when cross-building? I have Mac M1 so I need a newer cmake version too, but I'm not cross-building.

**Review by uilianries — COMMENTED** [review]
Thank you for your contribution, please, revert `build_requirements` methods changes, except for newer cmake version. pkgconfig should not be required, as CMakeDeps should be enough, also, installing cmake for cross-building only affects Mac M1 users for instance.

File: recipes/ccache/all/conanfile.py (hunk 9 above)
Comment (Cogitri) [reply recipes/ccache/all/conanfile.py]: Unfortunately that doesn't work, as per CI log.

Comment (uilianries) [reply recipes/ccache/all/conanfile.py]: I'll take a look locally this afternoon. Tahnk you for pinging

Comment (Cogitri) [reply recipes/ccache/all/conanfile.py]: Thanks :)

**Comment by uilianries** [comment]
@Cogitri please, consider https://github.com/dampsoft/conan-center-index/pull/1

**Review by uilianries — APPROVED** [review]
LGTM

**Review by jwillikers — APPROVED** [review]

**Comment by prince-chrismc** [comment]
https://github.com/conan-io/conan-center-index/pull/16801 To help with missing packages

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
Looks super good!

**Review by SSE4 — APPROVED** [review]

**Comment by prince-chrismc** [comment]
Let's try v2

**Pull request merged**